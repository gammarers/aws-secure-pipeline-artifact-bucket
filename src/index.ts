import { SecureBucket, SecureBucketProps } from '@gammarers/aws-secure-bucket';
import { Stack } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';


export interface SecurePipelineArtifactBucketProps extends SecureBucketProps {}

export class SecurePipelineArtifactBucket extends SecureBucket {

  constructor(scope: Construct, id: string, props?: SecurePipelineArtifactBucketProps) {
    super(scope, id, props);

    // ðŸ‘‡ Get account & region
    const account = Stack.of(this).account;
    const region = Stack.of(this).region;
    // ðŸ‘‡ Get qualifier
    const qualifier = Stack.of(this).synthesizer.bootstrapQualifier || 'hnb659fds';

    this.addToResourcePolicy(new iam.PolicyStatement({
      actions: [
        's3:*',
      ],
      resources: [
        `${this.bucketArn}`,
        `${this.bucketArn}/*`,
      ],
      principals: [
        new iam.ArnPrincipal(`arn:aws:iam::${account}:role/cdk-${qualifier}-deploy-role-${account}-${region}`),
      ],
    }));
  }

}
