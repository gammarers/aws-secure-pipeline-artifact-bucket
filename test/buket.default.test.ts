import { App, Stack } from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { SecurePipelineArtifactBucket } from '../src';

describe('SecurePipelineArtifactBucket default Testing', () => {

  const stack = new Stack(new App(), 'TestingStack', {
  });

  const bucket = new SecurePipelineArtifactBucket(stack, 'SecurePipelineArtifaceBucket');

  it('Is Bucket', () => {
    expect(bucket).toBeInstanceOf(s3.Bucket);
  });

  const template = Template.fromStack(stack);

  it('Should have encryption', () => {
    template.hasResourceProperties('AWS::S3::BucketPolicy', {
      Bucket: Match.objectEquals({
        Ref: Match.stringLikeRegexp('SecurePipelineArtifaceBucket'),
      }),
      PolicyDocument: {
        Version: '2012-10-17',
        Statement: Match.arrayWith([
          Match.objectEquals({
            Action: 's3:*',
            Effect: 'Allow',
            Principal: {
              AWS: {
                'Fn::Join': [
                  '',
                  [
                    'arn:aws:iam::',
                    {
                      Ref: 'AWS::AccountId',
                    },
                    ':role/cdk-hnb659fds-deploy-role-',
                    {
                      Ref: 'AWS::AccountId',
                    },
                    '-',
                    {
                      Ref: 'AWS::Region',
                    },
                  ],
                ],
              },
            },
            Resource: [
              {
                'Fn::GetAtt': [
                  Match.stringLikeRegexp('SecurePipelineArtifaceBucket'),
                  'Arn',
                ],
              },
              {
                'Fn::Join': [
                  '',
                  [
                    {
                      'Fn::GetAtt': [
                        Match.stringLikeRegexp('SecurePipelineArtifaceBucket'),
                        'Arn',
                      ],
                    },
                    '/*',
                  ],
                ],
              },
            ],
          }),
        ]),
      },
    });
  });


  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
