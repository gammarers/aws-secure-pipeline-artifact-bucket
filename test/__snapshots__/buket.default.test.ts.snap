// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurePipelineArtifactBucket default Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "SecurePipelineArtifaceBucket112332C8": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "Private",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "BucketOwnerEnforced",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecurePipelineArtifaceBucketPolicy2FF78204": {
      "Properties": {
        "Bucket": {
          "Ref": "SecurePipelineArtifaceBucket112332C8",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SecurePipelineArtifaceBucket112332C8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "SecurePipelineArtifaceBucket112332C8",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:aws:iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":role/cdk-hnb659fds-deploy-role-",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      "-",
                      {
                        "Ref": "AWS::Region",
                      },
                    ],
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SecurePipelineArtifaceBucket112332C8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "SecurePipelineArtifaceBucket112332C8",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
